lib/stripe/client_api_operations.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+module Stripe
+  # Define instance methods on the including class (i.e. StripeClient)
+  # to access API resources.
+  module ClientAPIOperations
+    # Proxy object to inject the client into API resources. When included,
+    # all resources are defined as singleton methods on the client in the
+    # plural form (e.g. Stripe::Account => client.accounts).
+    class ClientProxy
+      def initialize(client:, resource: nil)
+        @client = client
+        @resource = resource
+      end
+
+      attr_reader :client
+
+      def with_client(client)
+        @client = client
+        self
+      end
+
+      # Used to either send a method to the API resource or the nested
+      # ClientProxy when a resource is namespaced.
+      def method_missing(method, *args)
+        super unless @resource.respond_to?(method)
+
+        update_args_with_client!(method, args)
+
+        @resource.public_send(method, *args)
+      end
+
+      def respond_to_missing?(symbol, include_private = false)
+        super unless @resource
+        @resource.respond_to?(symbol) || super
+      end
+
+      # Since the method signature differs when operating on a collection versus
+      # a singular resource, it's required to perform introspection on the
+      # parameters to respect any passed in options or overrides.
+      #
+      # Two noteworthy caveats:
+      # 1) Does not merge into methods that use `_opts` as that means
+      #    the param is unused.
+      # 2) Preserves incorrect options (e.g. passing nil) so that APIResource
+      #    can handle errors.
+      def update_args_with_client!(method, args)
+        opts_pos = @resource.method(method).parameters.index(%i[opt opts])
+
+        return unless opts_pos
+
+        opts = opts_pos >= args.length ? {} : args[opts_pos]
+
+        normalized_opts = Stripe::Util.normalize_opts(opts)
+        args[opts_pos] = { client: @client }.merge(normalized_opts)
+      end
```

Comment:
The primary regressions that I found during the first PR all had to do with nuances when it came time to merge the options. I was able to identify the scenarios and lock them in under test: https://github.com/stripe/stripe-ruby/pull/954/files#diff-50b82fc5b424d8d4c21c2ddd6c67485e40fbc76fdf5cc1e4320949a4ad20fb62R56